lsp-types/src/Language/LSP/MetaModel/Types.hs

```diff
@@ -0,0 +1,225 @@
+{-# LANGUAGE GADTs                 #-}
+{-# LANGUAGE KindSignatures        #-}
+{-# LANGUAGE LambdaCase            #-}
+{-# LANGUAGE OverloadedStrings     #-}
+{-# LANGUAGE ScopedTypeVariables   #-}
+{-# LANGUAGE TemplateHaskell       #-}
+{-# LANGUAGE TypeApplications      #-}
+{-|
+This defines the types of the LSP "metamodel", which is a machine-readable format specifying the
+types used in the LSP protocol.
+
+The type system is quite typescript-y, which isn't surprising given that the whole protocol is
+very typescript-y.
+
+A typescript version of the metamodel types can be found here, which is useful for constructing
+this Haskell model of them:
+https://github.com/microsoft/vscode-languageserver-node/blob/main/tools/src/metaModel.ts
+-}
+module Language.LSP.MetaModel.Types where
+
+import           Data.Aeson         hiding (Null, String)
+import qualified Data.Aeson         as JSON
+import qualified Data.Aeson.TH      as JSON
+import qualified Data.Char          as Char
+import           Data.Text          (Text)
+
+import           Control.Lens
+import qualified Data.List.NonEmpty as NE
+
+data MessageDirection = ServerToClient | ClientToServer | Both
+  deriving (Show, Eq, Ord)
+
+instance ToJSON MessageDirection where
+  toJSON ServerToClient = toJSON @String "serverToClient"
+  toJSON ClientToServer = toJSON @String "clientToServer"
+  toJSON Both           = toJSON @String "both"
+
+instance FromJSON MessageDirection where
+  parseJSON = withText "MessageDirection" $ \case
+    "serverToClient" -> pure ServerToClient
+    "clientToServer" -> pure ClientToServer
+    "both"           -> pure Both
+    t                -> fail $ "unknown message direction " ++ show t
+
+data BaseTypeName = URI | DocumentUri | Integer | UInteger | Decimal | RegExp | String | Boolean | Null
+  deriving (Show, Eq, Ord)
+
+data Property = Property
+  { name          :: Text
+  , type_         :: Type
+  , optional      :: Maybe Bool
+  , documentation :: Maybe Text
+  , since         :: Maybe Text
+  , proposed      :: Maybe Bool
+  }
+  deriving (Show, Eq, Ord)
+
+data StructureLiteral = StructureLiteral
+  { properties    :: [Property]
+  , documentation :: Maybe Text
+  , since         :: Maybe Text
+  , proposed      :: Maybe Bool
+  }
+  deriving (Show, Eq, Ord)
+
+data Type =
+  BaseType { btName :: BaseTypeName }
+  | ReferenceType { rtName :: Text }
+  | ArrayType { atElement :: Type }
+  | MapType { mKey :: Type, mValue :: Type }
+  | AndType { aItems :: NE.NonEmpty Type }
+  | OrType { oItems :: NE.NonEmpty Type }
+  | TupleType { tItems :: [Type] }
+  | StructureLiteralType { stlValue :: StructureLiteral }
+  | StringLiteralType { slValue :: Text }
+  | IntegerLiteralType { ilValue :: Integer }
+  | BooleanLiteralType { blValue :: Bool }
+  deriving (Show, Eq, Ord)
+
+data Request = Request
+  { method              :: Text
+  , params              :: Maybe Type -- typescript says it can be [Type], but it never is so whatever
```

Comment:
I'd argue this should match the spec, better to match the spec and reduce churn later imo